### DISCUSSION TITLE
Лейблы

### DISCUSSION THREAD

**Discussion by tupizm** [Chimildic/YaMuTools#47] [Идея] [answered]
Прошу прощения, я снова со своими лейблами. Нельзя ли как автоматом собрать все альбомы одного лейбла. Допустим выбрать любой альбом лейбла, и из него получить название лейбла, ну и далее собрать уже альбомы. Вручную это вроде работает, но если альбомов много, то тяжеловато тыкать.

**Comment by Chimildic** [comment, marked as answer]
Вряд ли это имеет смысл для расширения YaMuTools. В одном лейбле может быть сотня другая альбомов. Толку от этой кучи треков никакого.

Самый простой и гибкий для вас вариант это [goofy](https://github.com/Chimildic/goofy) и импорт в Яндекс.

Можно поэкспериментировать в рамках Яндекс.Музыки через [gooex](https://github.com/Chimildic/gooex). У него нет _опрятных_ функций для этого дела. Потому что не вижу интереса к этому проекту. Но как пример:
```js
// gooex build 2022.04.16
function collectLabelTracks() {
  let albumId = '12246669';
  let album = gooex.Album.getAlbumWithTracks(albumId);
  let labelAlbums = getLabelAlbums(album.labels[0].id, 0, 'rating');
  let tracks = getAlbumsTracks(labelAlbums);

  function getLabelAlbums(labelId, page, sortBy) {
    return gooex.customRequest()
      .withPath(`/labels/${labelId}/albums`)
      .withQuery({ page: page, 'sort-by': sortBy})
      .get().albums;
  }

  function getAlbumsTracks(albums) {
    return albums.map(a => gooex.Album.getAlbumWithTracks(`${a.id}`).volumes).flat(2);
  } 
}

```

**Reply by LyokhaOhm** [reply]
> Вряд ли это имеет смысл для расширения YaMuTools
— Я считаю, что как раз таки очень имеет смысл именно для YaMuTools :)

> В одном лейбле может быть сотня другая альбомов.
— Именно поэтому нужен инструмент для работы с лейблами
В случае работы с независимыми и концептуальными лейблами это особенно ценный функционал.

> Толку от этой кучи треков никакого
— Если говорить именно о «куче», то да — нет никакого смысла.
Но если говорить об упорядоченной «дискографии» лейбла — то это именно то, что нужно.

Думаю нужен функционал, аналогичный тому как собираются плейлисты типа «Дискография исполнителя»

На странице лейблов на Яндекс Музыке альбомы отсортированы по хронологии их даты выхода.
Создание плейлиста «Дискография лейбла» — это собрание в один плейлист альбомов лейбла. 
Например, в порядке сортировки альбомов «по дате выхода» как их сортирует Яндекс Музыка на странице лейбла.

В итоге получать на выходе плейлист «Дискография Лейбла»
Внутри плейлиста сортировка треков:
– Альбомы «по дате выхода» (от старых к новым / от новых к старым)
–– но треки при этом всегда сортированы по их порядку в альбоме
то есть вне зависимости от того как будут сортированы «альбомы» в плейлисте, «треки» всегда должны быть сортированы в порядке их нумерации в альбоме (это нужно как раз чтобы не получилось «кучи»)

Тому, кому нужна будет «куча» — применят инструмент YaMuTools для перемешивания треков в плейлисте

**Comment by LyokhaOhm** [comment]
Поддерживаю! Очень не хватает расширенного инструмента работы с лэйблами.

**Reply by kisoid** [reply]
> Поддерживаю! Очень не хватает расширенного инструмента работы с лэйблами.

Тоже поддерживаю. Лучше иметь возможность, чем не иметь.
Лейблы бывают специализирующиеся на каких-то стилях.

> Можно поэкспериментировать в рамках Яндекс.Музыки через gooex

~~Либо нужен подробный мануал для тех, кто не шарит в JS, как воспользоваться gooex~~
Не, нафиг. Полистал https://chimildic.github.io/gooex/ - как-то всё очень заморочено...
Токен надо создавать, да ещё подписка требуется...

**Comment by Chimildic** [comment]
Далеко не все предложенное сделано. Но просто собрать альбомы в кучу получится. Попробуйте версию 0.8.3, доступна в Chrome Web Store.

**Reply by LyokhaOhm** [reply]
Отличная новость! Уже пробую... Сразу даю обратную связь:

— Думаю, что лучше сделать возможность при создании плейлиста указывать интервал лет, за период которых нужно собрать плейлист дискографии лейбла. (Сейчас при создании появляется запрос на указании количества альбомов — это не имеет практического смысла.)
Есть большой практический смысл указывать именно период, за который нужно собрать дискографию лейбла — за один год или за период лет, или декаду.

Есть смысл автоматически создавать плейлисты ⁠лейбла по годам — сразу несколько листов по годам.
То есть, чтобы YaMuTools по одному запросу сразу делал несколько плейлистов дискографии лейбла по годам — «один год = один плейлист» ... десять лет лейблу = десять плейлистов дискографии по годам

___
Но даже если просто собирать всю дискографии лейбла в один плейлист, как сделано сейчас — это уже большой огромный плюс YaMuTool

**Reply by LyokhaOhm** [reply]
То что вы сделали фильтры по типу — за это вам отдельное большое спасибо!
Вы превзошли ожидание :) Я даже не рассчитывал на такую возможность, но очень хорошо, что вы это сделали!

**Reply by LyokhaOhm** [reply]
Особое большое спасибо за возможность сделать отдельный плейлист из сборников лейбла!

**Reply by LyokhaOhm** [reply]
Попробовал протестировать на лейбле Warp (https://music.yandex.ru/label/40101)

По запросу на собрание всех релизов лейбла получил вот такой плейлист:
https://music.yandex.ru/users/LyokhaOhm/playlists/1878

Посмотрел начало и конец — похоже, вроде, нормально дискография собирается.
Но есть непонятный момент — в конце плейлиста непонятным образом вставлены 5 треков вразрез хронологии дискографии и не соответствуют никакой последовательности.

Проверьте, пожалуйста

**Reply by LyokhaOhm** [reply]
Сделал плейлист по запросу «только сборники»
https://music.yandex.ru/users/LyokhaOhm/playlists/1879
Вроде нормально собрался

Заметил, что YaMuTools предложил выбор:
_– создать новый плейлист
– обновить существующий плейлист
– дополнить существующий плейлист_
Это хорошо, что предусмотрели такой выбор. Спасибо.

Однако я выбрал «создать новый плейлист».
В итоге получил плейлист с названием «**Альбомы** лейбла Warp», 
хотя делал запрос на «все **сборники** лейбла»

**Reply by LyokhaOhm** [reply]
Предлагаю автоматически давать плейлистам следующие названия:

___
Название:
– Дискография лейбла <название_лейбла>
_(все релизы за всё время, максимально полный плейлист)_
и описание плейлиста:
_Все релизы лейбла на момент создания плейлиста.
Всего: <____>_

___
Название:
– Альбомы лейбла <название_лейбла>
_(все альбомы лейбла)_
и описание плейлиста:
_Все альбомы лейбла на момент создания плейлиста.
Всего: <____>_

___
Название:
– Синглы лейбла <название_лейбла>
_(все синглы лейбла)_
и описание плейлиста:
_Все синглы лейбла на момент создания плейлиста.
Всего: <____>_

___
Название:
– Сборники лейбла <название_лейбла>
_(все сборники лейбла)_
и описание плейлиста:
_Все сборники лейбла на момент создания плейлиста.
Всего: <____>_

**Reply by LyokhaOhm** [reply]
Итоги

Собрал тестовые плейлисты лейбла, на примере лейбла «Warp Records»
Warp (https://music.yandex.ru/label/40101)

___

https://music.yandex.ru/users/LyokhaOhm/playlists/1878
Название: 
**Дискография лейбла Warp**
Описание: 
Все **релизы** лейбла на момент создания плейлиста.
Всего: 678

https://music.yandex.ru/users/LyokhaOhm/playlists/1882
Название: 
**Альбомы лейбла Warp**
Описание: 
Все **альбомы** лейбла на момент создания плейлиста.
Всего: 498

https://music.yandex.ru/users/LyokhaOhm/playlists/1881
Название: 
**Синглы лейбла Warp**
Описание: 
Все **синглы** лейбла на момент создания плейлиста.
Всего: 157

https://music.yandex.ru/users/LyokhaOhm/playlists/1879
Название: 
**Сборники лейбла Warp**
Описание: 
Все **сборники** лейбла на момент создания плейлиста.
Всего: 23

___

**Пока обнаружил баги при создании плейлистов включающих альбомы**

Пример:
https://music.yandex.ru/users/LyokhaOhm/playlists/1878
здесь в самом конце 5 (пять) треков находятся не на своём месте

– https://music.yandex.ru/users/LyokhaOhm/playlists/1882
здесь в самом конце 4 (четыре) трека находятся не на своём месте

**Reply by LyokhaOhm** [reply]
Пришлось поменять ссылки.  
Если нужно, пришлю актуальные.

**Reply by Chimildic** [reply]
С описанием странно выходит. Далеко не каждое использование должно заканчиваться сбором "всех" релизов в один плейлист.

**Comment by LyokhaOhm** [comment]
@Chimildic нужны ли ещё подробности?

При создании плейлиста с дискографией лейбла [Warp](https://music.yandex.ru/label/40101) наблюдаю «аномальную» установку некоторых треков в конец плейлиста там, где их не должно быть

Пример на альбоме [Danny Brown – Atrocity Exhibition](https://music.yandex.ru/album/6372900) 
В плейлисте с полной дискографией лейбла все треки этого альбома располагаются скриптом на позициях по порядковому номеру в плейлисте `от 986 до 999` НО последний трек из этого альбома скрипт ставит в плейлисте на позицию `4542` 

_Это просто один конкретный пример из множества, чтобы более понятно стало, что я имею в виду_

**Reply by Chimildic** [reply]
Баг с позиционированием при вставке большего числа треков. Поправлю по случаю обновления.

**Comment by LyokhaOhm** [comment]
Пример 1 – создание **ПОЛНОГО** плейлиста дискографии лейбла [Warp](https://music.yandex.ru/label/40101)
Смотри [скриншот](https://prnt.sc/aGNjmQXfH-w2) — эти треки не должны быть в конце плейлиста

Пример 2 – создание плейлиста только с **АЛЬБОМАМИ** лейбла [Warp](https://music.yandex.ru/label/40101)
Смотри [скриншот](https://prnt.sc/SRYzTEQQrAc0) — эти треки не должны быть в конце плейлиста

**Comment by LyokhaOhm** [comment]
Не получается создать плейлист дискографии лейбла [Ninja Tune](https://music.yandex.ru/label/447329/albums?sort=year)

Процесс начинается, но не завершается

**Reply by Chimildic** [reply]
Не воспроизводится - плейлист создался, и минуты не прошло. Нужно больше деталей. 
Откройте консоль разработчика (Ctrl + Shift + J). Нужен скриншот с ошибками. Желательно перед добавлением очистить консоль (Ctrl + L). Потому что там много ошибок от Яндекса.
